### DISCUSSION TITLE
Which Evennia contribs are you using?

### DISCUSSION THREAD

**Discussion by Griatch** [evennia/evennia#2414] [General Discussion]
Evennia contains a slew of different contributions - optional code you can use/borrow/modify for your own purposes. Which ones are you using (or plan to use) and how? 

These are currently the contribs in master branch ([evennia/contribs/](https://github.com/evennia/evennia/tree/master/evennia/contrib)): 

## Contrib modules

* Barter system (Griatch 2012) - A safe and effective barter-system
  for any game. Allows safe trading of any goods (including coin).
* Building menu (vincent-lg 2018) - An `@edit` command for modifying
  objects using a generated menu. Customizable for different games.
* CharGen (Griatch 2011) - A simple Character creator for OOC mode.
  Meant as a starting point for a more fleshed-out system.
* Clothing (FlutterSprite 2017) - A layered clothing system with
  slots for different types of garments auto-showing in description.
* Color-markups (Griatch, 2017) - Alternative in-game color markups.
* Custom gametime (Griatch, vlgeoff 2017) - Implements Evennia's
  gametime module but for custom game world-specific calendars.
* Dice (Griatch 2012) - A fully featured dice rolling system.
* Email-login (Griatch 2012) - A variant of the standard login system
  that requires an email to login rather then just name+password.
* Evscaperoom (Griatch 2019) - A full engine for making escaperoom puzzles
* Extended Room (Griatch 2012) - An expanded Room typeclass with
  multiple descriptions for time and season as well as details.
* Field Fill (FlutterSprite 2018) - A simple system for creating an
  EvMenu that presents a player with a highly customizable fillable
  form
* GenderSub (Griatch 2015) - Simple example (only) of storing gender
  on a character and access it in an emote with a custom marker.
* Health Bar (Tim Ashley Jenkins 2017) - Tool to create colorful bars/meters.
* Mail (grungies1138 2016) - An in-game mail system for communication.
* Menu login (Griatch 2011, 2019, Vincent-lg 2016) - A login system using menus asking
  for name/password rather than giving them as one command.
* Map Builder (CloudKeeper 2016) - Build a game area based on a 2D
  "graphical" unicode map. Supports assymmetric exits.
* Menu Login (Vincent-lg 2016) - Alternate login system using EvMenu.
* Multidescer (Griatch 2016) - Advanced descriptions combined from
  many separate description components, inspired by MUSH.
* Puzzles (Hendher 2019) - Combine objects to create new items, adventure-game style
* Random String Generator (Vincent Le Goff 2017) - Simple pseudo-random
  generator of strings with rules, avoiding repetitions.
* RPLanguage (Griatch 2015) - Dynamic obfuscation of emotes when
  speaking unfamiliar languages. Also obfuscates whispers.
* RPSystem (Griatch 2015) - Full director-style emoting system
  replacing names with sdescs/recogs. Supports wearing masks.
* Security/Auditing (Johhny 2018) - Log server input/output for debug/security.
* Simple Door - Example of an exit that can be opened and closed.
* Slow exit (Griatch 2014) - Custom Exit class that takes different
  time to pass depending on if you are walking/running etc.
* Talking NPC (Griatch 2011) - A talking NPC object that offers a
  menu-driven conversation tree.
* Tree Select (FlutterSprite 2017) - A simple system for creating a
  branching EvMenu with selection options sourced from a single
  multi-line string.
* Wilderness (titeuf87 2017) - Make infinitely large wilderness areas
  with dynamically created locations.
* UnixCommand (Vincent Le Geoff 2017) - Add commands with UNIX-style syntax.

## Contrib packages

* In-game Python (Vincent Le Goff 2017) - Allow trusted builders to script
  objects and events using Python from in-game.
* Turnbattle (FlutterSprite 2017) - A turn-based combat engine meant
  as a start to build from. Has attack/disengage and turn timeouts,
  and includes optional expansions for equipment and combat movement, magic 
  and ranged combat.
* Tutorial examples (Griatch 2011, 2015) - A folder of basic
  example objects, commands and scripts.
* Tutorial world (Griatch 2011, 2015) - A folder containing the
  rooms, objects and commands for building the Tutorial world.

## Upcoming in `develop` branch (Evennia 1.0):
(tentative list)
- Crafting system (Griatch, 2020) - Full, extendable crafting system.
- Traits (Whitenoise 2015, Griatch 2021) - Properties for handling and tracking changing RPG skills and stat values.

**Comment by kovitikus** [comment]
Smooth-brains like me prefer to reinvent the wheel. 

There's a lot of great stuff in the contribs, and plenty of folk keep reminding me of it, but because I am too lazy to read them I create more work for myself. Don't be a smooth-brain.

Shame! 🔔 Shame! 🔔 Shame! 🔔

**Comment by davewiththenicehat** [comment]
clothing, modded it to work off of my Character's body parts and created a armor type class for each race that inherits clothing.
gender sub. I also recall it having odd messages to caller.
rpsystem, I wound up building a custom emote to turn "/target" into get_display_name. Than added '/Me' and '/Target' capitalizing the first letter without lower casing the rest of the string.

I plan to use rplanguage, gametime, email login and healthbar.
Most of the others I plan to use parts or methods within.

I tried traits and wound up making an iteratable object. It functions as a dictionary that forwards to the database. I use it for skills, body parts, object condition, damage reduction, damage types so on. I don't recall the specifics of why I chose not to use traits.

So far rpsystem is my favorite. I really really like that contrib.

**Comment by strikaco** [comment]
#### Which ones are you using (or plan to use) and how?
>     * Dice (Griatch 2012) - A fully featured dice rolling system.
>     * GenderSub (Griatch 2015) - Simple example (only) of storing gender
>       on a character and access it in an emote with a custom marker.
>     * Health Bar (Tim Ashley Jenkins 2017) - Tool to create colorful bars/meters.
>     * Mail (grungies1138 2016) - An in-game mail system for communication.
>     * UnixCommand (Vincent Le Geoff 2017) - Add commands with UNIX-style syntax.

These are all basic features that I'm using as-is, with the exception of UnixCommand-- I extended that to dynamically define its parsers based on the structure of any provided Django form; the result lets one complete/render forms from the CLI.

**Reply by ghost** [reply]
I'm technically using the mail contrib but I've rewritten the entire thing a few times.

The rest really aren't compatible with the free-form RPG style game I am trying to make, and those that are, I would rather do myself so they mesh well with the rest of my systems.

-Darren Wheeler

**Comment by pruuk** [comment]
I'm using:

- A modified version of the Building Menu
- Traits
- Custom Gametime (I went so far down the rabbit hole I have orbits modeled for my entire system)
- a modified version of the Healthbar
- Map Builder
- I also borrowed heavily from Ainneve's equipment module

I plan on using:

- A modified version of the clothing module to describe character mutations, such as armored scales
- Extended Room
